pkg/apply_gen/main.go

```diff
@@ -21,60 +21,43 @@ import (
 	networkingv1 "k8s.io/api/networking/v1"
 	policyv1beta1 "k8s.io/api/policy/v1beta1"
 	rbacv1 "k8s.io/api/rbac/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/client-go/rest"
 )
 
-func annotate(meta *metav1.ObjectMeta, rev int64, data []byte) {
+type applyParams struct {
+	isNamespaced   bool
+	forceConflicts bool
+}
+
+func annotate(meta *metav1.ObjectMeta, rev int64) {
 	if meta.Annotations == nil {
 		meta.Annotations = make(map[string]string)
 	}
 	meta.Annotations[AnnotationResourceRevision] = strconv.FormatInt(rev, 10)
-	meta.Annotations[AnnotationResourceOriginal] = string(data)
 }
 {{- range . }}
 
-func apply{{ .Kind }}(o *{{ .API }}.{{ .Kind }}, data []byte, rev int64, getFunc func(string, metav1.GetOptions) (*{{ .API }}.{{ .Kind }}, error), createFunc func(*{{ .API }}.{{ .Kind }}) (*{{ .API }}.{{ .Kind }}, error), patchFunc func(string, types.PatchType, []byte, ...string) (*{{ .API }}.{{ .Kind }}, error)) error {
-	annotate(&o.ObjectMeta, rev, data)
-	current, err := getFunc(o.Name, metav1.GetOptions{})
-	if errors.IsNotFound(err) {
-		_, err = createFunc(o)
-		return err
-	}
-	if err != nil {
-		return err
-	}
-
+func apply{{ .Kind }}(o *{{ .API }}.{{ .Kind }}, rev int64, client rest.Interface, params *applyParams) error {
```

Comment:
The motivation when `apply` was first implemented as template-rendering was
to use typed REST client for safer operations.  Now that `rest.Interface`, a non-typed
REST client, is used, there is no reasons to use templates.

Would you please implement this just as a pure Go function?

Reply:
Agree with you, we will refactor this code in the next PR of this task.